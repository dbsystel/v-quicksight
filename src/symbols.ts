export const EmbeddingContextInjectionKey = Symbol(
  "The key to be used with inject when accessing the created embeddingContext for usage with the amazon-quicksight-embedding-sdk",
)
