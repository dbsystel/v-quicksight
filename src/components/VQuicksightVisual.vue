<script setup lang="ts">
import type { VQuicksightFrameOptions, VQuicksightVisualContentOptions } from '../types'
import type {
  EmbeddingContext,
  ExperienceFrameMetadata,
  FrameOptions,
  Parameter,
  EmbeddingEvents,
  VisualContentOptions,
  VisualExperience
} from 'amazon-quicksight-embedding-sdk'
import { nanoid } from 'nanoid'
import type { Ref } from 'vue'
import { computed, inject, ref, watch } from 'vue'
import { EmbeddingContextInjectionKey } from '../symbols'

const props = withDefaults(
  defineProps<
    VQuicksightFrameOptions &
      VQuicksightVisualContentOptions & {
        container?: string | HTMLElement // override from FrameOptions and make it optional
        id?: string
      }
  >(),
  {
    width: '100%',
    height: '100%',
    resizeHeightOnSizeChangedEvent: true,
    fitToIframeWidth: true,
    parameters: () => []
  }
)

const emit = defineEmits<{
  (e: 'change', data: { changeEvent: EmbeddingEvents; metadata?: ExperienceFrameMetadata }): void
  (
    e: 'message',
    data: { messageEvent: EmbeddingEvents; experienceMetadata?: ExperienceFrameMetadata }
  ): void
}>()

const embeddingContext = inject<Ref<EmbeddingContext>>(EmbeddingContextInjectionKey)

const visualFrame = ref<VisualExperience>()

const containerId = computed(() => props.id || `v-quicksight-visual-${nanoid(6)}`)
const frameOptions = computed<FrameOptions>(() => {
  return {
    url: props.url,
    container: props.container || `#${containerId.value}`,
    height: props.height,
    width: props.width,
    className: props.className,
    resizeHeightOnSizeChangedEvent: props.resizeHeightOnSizeChangedEvent,
    withIframePlaceholder: props.withIframePlaceholder,
    onChange: (changeEvent, metadata) => {
      emit('change', { changeEvent, metadata })
    }
  }
})
const contentOptions = computed<VisualContentOptions>(() => {
  return {
    parameters: props.parameters,
    locale: props.locale,
    fitToIframeWidth: props.fitToIframeWidth,
    onMessage: async (messageEvent, experienceMetadata) => {
      emit('message', { messageEvent, experienceMetadata })
    }
  }
})

async function embedVisual(ctx: EmbeddingContext) {
  visualFrame.value = await ctx.embedVisual(frameOptions.value, contentOptions.value)
}

async function setParameters(frame: VisualExperience, parameters: Parameter[]) {
  return await frame.setParameters(parameters)
}

watch(
  () => embeddingContext?.value,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      embedVisual(newValue)
    }
  },
  { immediate: true }
)

watch(
  () => props.parameters,
  (newValue) => {
    if (newValue && visualFrame.value) {
      setParameters(visualFrame.value, newValue)
    } else if (embeddingContext?.value) {
      embedVisual(embeddingContext?.value)
    } else {
      console.error('could not find any embeddingContext for updating parameters')
    }
  },
  { deep: true }
)
</script>

<template>
  <div class="v-quicksight-visual" :id="containerId"></div>
</template>
