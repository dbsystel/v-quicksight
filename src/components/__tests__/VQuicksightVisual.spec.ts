import VQuicksightVisual from '../VQuicksightVisual.vue'
import { EmbeddingContextInjectionKey } from '../../symbols'
import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { nextTick, ref } from 'vue'

const frameOptions = {
  container: '#custom-el',
  width: '300px',
  height: '700px',
  resizeHeightOnSizeChangedEvent: false,
  withIframePlaceholder: true,
  className: 'my-custom-class'
}
const contentOptions = {
  parameters: [
    {
      Name: 'minimum',
      Values: [10]
    }
  ],
  fitToIframeWidth: false,
  locale: 'en-US'
}

describe('VQuicksightVisual', () => {
  function setupComponent(props: Record<string, any> = {}) {
    const visualFrame = {
      setParameters: vi.fn().mockResolvedValue({ success: true })
    }
    const embeddingContext = {
      embedVisual: vi.fn().mockResolvedValue(visualFrame)
    }

    const component = mount(VQuicksightVisual, {
      global: {
        provide: {
          [EmbeddingContextInjectionKey]: ref(embeddingContext)
        }
      },
      props: {
        ...props,
        url: 'https://my-embed-url.example.org',
        id: 'v-quicksight-visual-1'
      }
    })
    return { component, visualFrame, embeddingContext }
  }

  it('renders as expected', () => {
    const { component } = setupComponent()
    expect(component.element).toMatchSnapshot()
  })

  it('should call embedVisual with default parameters', async () => {
    const { embeddingContext } = setupComponent()
    await nextTick()
    expect(embeddingContext.embedVisual).toHaveBeenCalledOnce()
    expect(embeddingContext.embedVisual).toHaveBeenCalledWith(
      {
        className: undefined,
        container: '#v-quicksight-visual-1',
        height: '100%',
        onChange: expect.anything(),
        resizeHeightOnSizeChangedEvent: true,
        url: 'https://my-embed-url.example.org',
        width: '100%',
        withIframePlaceholder: false
      },
      {
        locale: undefined,
        onMessage: expect.anything(),
        fitToIframeWidth: true,
        parameters: []
      }
    )
  })

  it('should call embedVisual with customized parameters', () => {
    const { embeddingContext } = setupComponent({ ...frameOptions, ...contentOptions })
    expect(embeddingContext.embedVisual).toHaveBeenCalledOnce()
    expect(embeddingContext.embedVisual).toHaveBeenCalledWith(
      {
        ...frameOptions,
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org'
      },
      {
        ...contentOptions,
        onMessage: expect.anything()
      }
    )
  })

  it('should call setParameters when the passed parameters are changing', async () => {
    const { visualFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({ url: '', parameters: contentOptions.parameters })
    expect(visualFrame.setParameters).toHaveBeenCalledOnce()
    expect(visualFrame.setParameters).toHaveBeenCalledWith(contentOptions.parameters)
  })
})
