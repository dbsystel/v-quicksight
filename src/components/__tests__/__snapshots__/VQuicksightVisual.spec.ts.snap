// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VQuicksightVisual > renders as expected 1`] = `
<div
  class="v-quicksight-visual"
  id="v-quicksight-visual-1"
/>
`;
